image: serialize RegistryImageReferences when checking signatures

haircommander opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

<!--
/kind api-change
/kind bug
/kind ci
/kind cleanup
/kind dependency-change
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake
/kind other
-->
/kind bug
#### What this PR does / why we need it:
So we can check a signature of an image that has both a tag and a digest
#### Which issue(s) this PR fixes:
fixes https://github.com/cri-o/cri-o/issues/8603
<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Fix a bug where signature checking failed if an image specified both a tag and a digest
```

haircommander commented on internal/storage/references/registry_reference.go:
```diff
@@ -36,6 +36,24 @@ func RegistryImageReferenceFromRaw(rawNamed reference.Named) RegistryImageRefere
+// Such image references are supported by docker but, due to their ambiguity, explicitly not by containers/image.
+//
+// This should only be called from internal/storage.
+func RegistryImageReferenceFromRawSerialized(rawNamed reference.Named) (RegistryImageReference, error) {
```
not crazy about this name, @mtrmac do you have any suggestions?

mtrmac replied:
See the other comment for a bit higher-level view…

… which still leaves open the naming of this canonicalization. My starting point is `canonicalize{TagDigest,DockerLike}` somewhere in the name, but that might get awfully long.

haircommander commented:
/retest

mtrmac commented on test/policy.bats:
```diff
@@ -204,3 +204,38 @@ SANDBOX_CONFIG="$TESTDATA/sandbox_config.json"
 	[[ "$output" == *"SignatureValidationFailed"* ]]
 }
+
+@test "allow signed image with restrictive policy on container creation7 if already pulled (by tag and ID)" {
```
This is fundamentally confusing things. “Image ID” (`StorageImageID`) is, historically (nowadays not always) the sha256 digest of the config, and usually expressed without the leading `sha256:`.

“image digest” is the sha256 of the manifest. The two are ~never the same for a single image.

In `busybox:latest@sha256:…`, the second value is the manifest digest. So assuming the value from `crictl images -q` is an image ID, using it in the manifest digest syntax should never find a match; and the test name is misleading.

(Also for the other added test.)

mtrmac reviewed (commented):
ACK to the overall functionality this achieves.

---

WRT the implementation … my question is “what rules are implied by the `RegistryImageReference` data type”? I think it should be one of:
- :tag@digest references are unambiguously allowed by the data type.
  - In that case the mapping of :tag@digest to @digest-only must be in consumers of `RegistryImageReference`, i.e. in `PullImage`/`IsRunningImageAllowed` (and what value should be used in `MetricImagePullsSuccessesInc`, with or without the tag?)
  - or maybe `RegistryImageReference` should have a `RawCanonicalized` method that returns a value suitable for c/image, and most callers of `Raw` should migrate
- :tag@digest references are unambiguously invalid in the data type. In that case
  - _the constructor of `RegistryImageReferenceFromRaw` should reject that value to enforce the rule_
  - It would be consistent for `ParseRegistryImageReferenceFromOutOfProcessData` to reject the data as well
  - We’d probably need some kind of `…FromRawCanonicalizing` and `Parse…CanonicalizingFromOutOfProcesss…` helpers, so that top-level callers don’t manipulate values directly.

The current PR is ambiguous about the permissibility of such values — the text parser canonicalizes them, but `…FromRaw` does not.

I weakly prefer the variant where :tag@digest values are invalid in `RegistryImageReference`, forcing its users to explicitly choose, by the constructor they call, whether they want to accept such values or not. If nothing else, the `parseImageParent`/`parseImageChild` RPC should never generate such values, and it can use the stricter parser/constructor.

haircommander commented:
I agree the library should take care of canonicalizing. In fact, I just replaced *FromRaw with this behavior. Also i believe I updated the test to test digest and not storage ID. PTAL again @mtrmac

mtrmac commented on internal/storage/references/registry_reference.go:
```diff
@@ -24,16 +24,27 @@ type RegistryImageReference struct {
+	if isTagged && isDigested {
+		canonical, err := reference.WithDigest(reference.TrimNamed(rawNamed), canonical.Digest())
+		if err != nil {
+			return RegistryImageReference{}, err
```
(The only way this can fail is if `canonical.Digest()` is invalid, and that shouldn’t have happened by induction, so I’d be fine with a `panic` here. But returning an error explicitly is perfectly valid as well.)

haircommander replied:
that's better because then I don't need to change the function signature

mtrmac replied:
Also using `panic` does not leave around (supposedly) 100% unreachable error handling code paths, considering that CRI-O is measuring code coverage.

mtrmac commented on internal/storage/references/registry_reference.go:
```diff
@@ -48,7 +59,7 @@ func ParseRegistryImageReferenceFromOutOfProcessData(input string) (RegistryImag
 	}
 	ref = reference.TagNameOnly(ref)
-	return RegistryImageReferenceFromRaw(ref), nil
+	return RegistryImageReferenceFromRaw(ref)
```
The documentation of `ParseRegistryImageReferenceFromOutOfProcessData` should document that it removes the tag from tag@digest inputs.

mtrmac commented on internal/storage/image.go:
```diff
@@ -966,21 +972,10 @@ func (svc *imageService) CandidatesForPotentiallyShortImageName(systemContext *t
-		// Strip the tag from ambiguous image references that have a
-		// digest as well (e.g.  `image:tag@sha256:123...`).  Such
-		// image references are supported by docker but, due to their
-		// ambiguity, explicitly not by containers/image.
```
(Absolutely non-blocking: Consider preserving some form of this comment, to have a documentation for the reason we modify the inputs.)

mtrmac reviewed (commented):
I’d prefer two variants of constructors, one only in the “XOR” form, one canonicalizing — but this works, up to you.

Assuming this design, code LGTM, just some documentation suggestions.

mtrmac commented on internal/storage/references/registry_reference.go:
```diff
@@ -40,6 +54,8 @@ func RegistryImageReferenceFromRaw(rawNamed reference.Named) RegistryImageRefere
 // It is only intended for communication with OUT-OF-PROCESS APIs,
 // like registry references provided by CRI by Kubelet.
+// It will modify the reference if both digest and tag are specified, stripping the tag and leaving the digest.
+// It will also verifies the image is not only a name. If it is only a name, the function errors.
```
The function does not fail in such a situation: The `TagNameOnly` call turns name-only references into _name_`:latest`

mtrmac reviewed (commented):
Do you also want to fix the linter-highlighted typo?

LGTM either way.

haircommander commented:
whoops, fixed!

mtrmac reviewed (commented):
/lgtm

haircommander commented:
/cherry-pick release-1.31

kwilczynski commented:
/approve
/lgtm

haircommander commented:
/retest

haircommander commented:
/retest

haircommander commented:
/retest
